src/ai/backend/manager/models/utils.py

```diff
@@ -63,79 +66,140 @@ def __init__(self, *args, **kwargs) -> None:
         self._readonly_txn_count = 0
         self._generic_txn_count = 0
         self._sess_factory = sessionmaker(self, expire_on_commit=False, class_=SASession)
+        self._readonly_sess_factory = sessionmaker(self, class_=SASession)
+
+    def _check_generic_txn_cnt(self) -> None:
+        if (
+            self._txn_concurrency_threshold > 0
+            and self._generic_txn_count >= self._txn_concurrency_threshold
+        ):
+            log.warning(
+                "The number of concurrent generic transactions ({}) "
+                "looks too high (warning threshold: {}).",
+                self._generic_txn_count,
+                self._txn_concurrency_threshold,
+                stack_info=False,
+            )
+
+    def _check_readonly_txn_cnt(self) -> None:
+        if (
+            self._txn_concurrency_threshold > 0
+            and self._readonly_txn_count >= self._txn_concurrency_threshold
+        ):
+            log.warning(
+                "The number of concurrent read-only transactions ({}) "
+                "looks too high (warning threshold: {}).",
+                self._readonly_txn_count,
+                self._txn_concurrency_threshold,
+                stack_info=False,
+            )
 
     @actxmgr
-    async def begin(self) -> AsyncIterator[SAConnection]:
-        async with super().begin() as conn:
+    async def _begin(self, connection: SAConnection) -> AsyncIterator[SAConnection]:
+        """
+        Begin generic transaction within the given connection.
+        """
+        async with connection.begin():
             self._generic_txn_count += 1
-            if (
-                self._txn_concurrency_threshold > 0
-                and self._generic_txn_count >= self._txn_concurrency_threshold
-            ):
-                log.warning(
-                    "The number of concurrent generic transactions ({}) "
-                    "looks too high (warning threshold: {}).",
-                    self._generic_txn_count,
-                    self._txn_concurrency_threshold,
-                    stack_info=False,
-                )
+            self._check_generic_txn_cnt()
             try:
-                yield conn
+                async with asyncio.timeout(self.conn_timeout):
```

Comment:
According to the usage of `connection_timeout`, it seems that it is **not actually a connection timeout** because it _does not_ guard the actual database connection establishment inside connection pools. The connection establishment is done at the outer outer `async with engine.connect()` block. Moreover, by definition, a transaction includes the beginning and conclusion of transactions, meaning that it should also include the outer `async with conn.begin()` block if you want to set a transaction timeout.

What is the exact meaning of this timeout and what is it guarding?
For me, it just seems to enforcing timeouts of the client-side database interfacing and data processing _inside_ transactions, _neither transactions nor connections_.
Shouldn't it include the transaction starts and shutdowns, i.e., the outer `async with conn.begin()` block, at least?

Reply:
I removed implementation of `conn_timeout` and I will cover this in another PR because it is not related to handling lifecycle of SQLAlchemy connection management.
Plus, I forgot why I made this `conn_timeout`.